### PR TITLE
md-icon: implementation 

### PR DESCRIPTION
**Pull request by programmist**
Ready for review.  Basic unit tests added, but more to come.

### CODE REVIEW

**Comment by ThomasBurleson** [comment]
@programmist - any chance you can squash the commit history?

**Comment by programmist** [comment]
Rebased with upstream master and squashed commits.

**Comment by ThomasBurleson** [comment]
@programmist - #woot

**Comment by programmist** [comment]
@ThomasBurleson I removed some of the unnecessary files from the demo and also shortened the SVG icon file sets in the demo.

File: src/components/icon/demoBasicUsage/index.html [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    </md-radio-group>
+    <md-input-container flex>
+      <label>Color</label>
+      <input ng-model="color">
+    </md-input-container>
+
+    <div class="glyph" ng-repeat="icon in icons | limitTo:limit" ng-if="iconType == 'SVG'">
+      <div class="preview-glyphs">
+        <md-icon ng-repeat="size in sizes" svg-icon="{{ icon.name }}"
+                 style="fill:{{ color }};width:{{ size }}px;height:{{ size }}px;" alt="{{ icon.name + size }}">
```
Comment (shyndman) [inline src/components/icon/demoBasicUsage/index.html]: Not super critical, but it seems like ng-style is appropriate here.

File: src/components/icon/icon.js [hunk 2]
```diff
@@ -20,25 +21,215 @@ angular.module('material.components.icon', [
...
+ *   $mdIcon('work:chair').then(function(chair) {
+ *     element.append(chair);
+ *   });
+ * };
+ * </hljs>
+ */
+function MdIconService(config, $http, $q, $log, $templateCache) {
+
+  var iconCache = {};
+  const urlRegex = /[-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?/gi;
```
Comment (shyndman) [inline src/components/icon/icon.js]: Is the const keyword valid across browsers?

Comment (gustavohenke) [inline src/components/icon/icon.js]: It's ES6: http://kangax.github.io/compat-table/es6/#const

Comment (shyndman) [inline src/components/icon/icon.js]: Right, that's my point.

File: src/components/icon/demoBasicUsage/index.html [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        </md-icon>
+      </div>
+      <div class="preview-scale">
+        <span class="step" ng-repeat="size in sizes">{{ size }}</span>
+      </div>
+      <div class="usage">
+        <input class="class" type="text" readonly="readonly" onclick="this.select();"
```
Comment (marcysutton) [inline src/components/icon/demoBasicUsage/index.html]: This input needs a working `<label>` that pairs them together.

File: src/components/icon/demoBasicUsage/style.css [hunk 4]
```diff
@@ -0,0 +1,112 @@
+* {
+  -moz-box-sizing: border-box;
+  -webkit-box-sizing: border-box;
+  box-sizing: border-box;
+  margin: 0;
+  padding: 0;
+}
+
+body {
```
Comment (marcysutton) [inline src/components/icon/demoBasicUsage/style.css]: The demo CSS shouldn't need this much boilerplate, no?

File: src/components/icon/icon.js [hunk 5]
```diff
@@ -20,25 +21,215 @@ angular.module('material.components.icon', [
...
+    return attr.fontIcon ? '<span class="md-font {{ fontIcon }}" role="img"></span>' : '';
+  }
+
+  /**
+   * Directive postLink
+   * Supports embeded SVGs, font-icons, & external SVGs
+   */
+  function postLink(scope, element, attr) {
+    var ariaLabel = attr.alt || scope.fontIcon || scope.svgIcon || scope.svgSrc;
+    $mdAria.expect(element, 'aria-label', ariaLabel);
```
Comment (marcysutton) [inline src/components/icon/icon.js]: Woot! Accessibility high five!

File: src/components/icon/icon.js (hunk 2 above)
Comment (ilanbiala) [inline src/components/icon/icon.js]: Not supported in IE10, please change it to `var`.

Comment (ThomasBurleson) [inline src/components/icon/icon.js]: Not true actually.  **const** has basic support for all modern browsers (except IE10 - which I did not know).
So since IE10 is not supported, we will need to use `var`.

File: src/components/icon/demoBasicUsage/index.html (hunk 1 above)
Comment (ilanbiala) [inline src/components/icon/demoBasicUsage/index.html]: `ng-style` is indeed appropriate.

Comment (ThomasBurleson) [inline src/components/icon/demoBasicUsage/index.html]: @ilanbiala - as mentioned before, the Angular Material team would like to constrain comments and reviews to ONLY code that is master or has been released. This code is not yet in master. Please wait until the team does its own internal views and changes and commits to master.

**Comment by marcysutton** [comment]
For labeling, I still think it would be cool to add alt text support similar to Polymer's `core-icon`. http://polymer.github.io/core-icon/components/core-icon/

Some examples would look like this:

**Example 1: intentionally leaving alt blank for a decorative image**
```html
<md-button>
   More 
   <md-icon src="icon-name" alt=""></md-icon>
</md-button>
```
Becomes:
```html
<md-button>
   <button>
       More 
       <!-- span hidden from accessibility layer with ARIA -->
       <md-icon src="icon-name"><span aria-hidden="true"></span></md-icon>
    </button>
</md-button>
```
**Example 2: alt text is provided in attribute**
```html
<md-icon alt="alt text provided"></md-icon>
```
Becomes:
```html
<md-icon src="icon-name">
   <span role="img" aria-label="alt text provided"></span>
</md-icon>
```

**Example 3: fallback, copies icon name as alt text**
```html
<md-icon src="icon-name"></md-icon>
```
Becomes:
```html
<md-icon>
   <span role="img" aria-label="icon-name"></span>
</md-icon>
```

File: src/components/icon/demoBasicUsage/index.html (hunk 1 above)
Comment (ilanbiala) [inline src/components/icon/demoBasicUsage/index.html]: @ThomasBurleson OSS is all about everyone contributing at every point. I don't know why you wouldn't want suggestions and feedback from everyone before code has released...that prevents bugs like what would have happened if you released this code with `const`. I'm just saying that you guys really need to let down on the restrictions and stuff and start just letting everyone contribute at any point and encouraging that.

File: src/components/icon/demoBasicUsage/index.html (hunk 3 above)
Comment (programmist) [inline src/components/icon/demoBasicUsage/index.html]: Fixed.

**Comment by programmist** [comment]
@marcysutton I updated the component to behave the way you described (The way polymer core-icon works).

**Comment by programmist** [comment]
@marcysutton Updated component alt text to work like Polymer core-icon.

All other comments addressed and fixes applied, if applicable.  Apparently when I squashed the commit it removed

File: src/components/icon/demoBasicUsage/style.css (hunk 4 above)
Comment (programmist) [inline src/components/icon/demoBasicUsage/style.css]: I removed as much of the CSS as I think I can.  Let me know if there's more I can do.

File: src/components/icon/icon.js (hunk 2 above)
Comment (programmist) [inline src/components/icon/icon.js]: Done.  I didn't know we were supporting IE < 11.

File: src/components/icon/demoBasicUsage/index.html (hunk 1 above)
Comment (programmist) [inline src/components/icon/demoBasicUsage/index.html]: Done.

**Comment by ThomasBurleson** [comment]
@programmist - things are looking good **except** we still need to separate the demos:

*  demo1 - show how font-icons are registered, used, and colored
*  demo2 - show how svg iconsets are registered and used with color
*  demo3 - show how external, individual svgs are loaded/used

> While your single demo does most of this, we had discussed the need for better granularity and use in the online demos docs.

**Comment by ThomasBurleson** [comment]
@marcysutton - really nice work with the **alt** suggestions and samples above.

File: docs/gulpfile.js [hunk 6]
```diff
@@ -108,8 +108,9 @@ module.exports = function(gulp, IS_RELEASE_BUILD) {
   gulp.task('docs-js', ['docs-app', 'docs-html2js', 'demos', 'build'], function() {
     return gulp.src([
       'bower_components/angularytics/dist/angularytics.js',
+      'bower_components/hammerjs/hammer.js',
       'dist/angular-material.js',
-      'dist/docs/js/**/*.js'
+      'dist/docs/js/**/*.js',
     ])
```
Comment (rschmukler) [inline docs/gulpfile.js]: I think we no longer need hammerjs, unless it was added back in? @programmist ? Also, trailing comma

File: gulpfile.js [hunk 7]
```diff
@@ -179,7 +179,13 @@ gulp.task('karma-sauce', function(done) {
  *
  ** ***************************************** */
 
-gulp.task('build', ['build-scss', 'build-js']);
+gulp.task('build', ['build-resources', 'build-scss', 'build-js']);
+
+gulp.task('build-resources', function() {
+  return gulp.src(['material-font/*'])
+    //.pipe(require('gulp-debug')())
```
Comment (rschmukler) [inline gulpfile.js]: Cleanup comment

File: src/components/icon/icon.js [hunk 8]
```diff
@@ -20,25 +21,221 @@ angular.module('material.components.icon', [
...
+ * </hljs>
+ */
+function MdIconService(config, $http, $q, $log, $templateCache) {
+
+  var iconCache = {};
+  var urlRegex = /[-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?/gi;
+  return function getIcon(id) {
+    id = id || '';
+
+    if (id.match(urlRegex)) {
```
Comment (rschmukler) [inline src/components/icon/icon.js]: `urlRegex.test(id)` is more performant here because it doesn't actually separate the match into the various match groups. Not really something that needs to be changed, just pointing it out.

File: src/components/icon/icon.js [hunk 9]
```diff
@@ -20,25 +21,221 @@ angular.module('material.components.icon', [
...
+  function getTemplate(element, attr) {
+    return attr.fontIcon ? '<span class="md-font" ng-class="fontIcon"></span>' : '';
+  }
+
+  /**
+   * Directive postLink
+   * Supports embeded SVGs, font-icons, & external SVGs
+   */
+  function postLink(scope, element, attr) {
+    if (attr.alt == '') {
```
Comment (rschmukler) [inline src/components/icon/icon.js]: We might want to use `scope.$watch` or `attr.$observe` here if we want to allow someone to have dynamic icons...

For example, I could see the following being a common use case:

```
<md-icon svg-icon="{{ selected ? 'checkbox-checked' : 'checkbox-unchecked' }}">
```

To help maintain performance we could register only one listener (depending on which mode they are using (`svgSrc`, `svgIcon` or `fontIcon`)

File: src/components/icon/icon.scss [hunk 10]
```diff
@@ -1,24 +1,3772 @@
...
+  font-style:normal;
+}
+
+@font-face {
+  font-family:"material";
+  src:url("/dist/material-font/material.eot");
+  src:url("/dist/material-font/material.eot?#iefix") format("embedded-opentype"),
+    url("/dist/material-font/material.woff") format("woff"),
+    url("/dist/material-font/material.ttf") format("truetype"),
+    url("/dist/material-font/material.svg?#material") format("svg");
```
Comment (rschmukler) [inline src/components/icon/icon.scss]: We will want to make the `font-face` root directory configurable and then inject it into a style tag during run time so that we don't force our users to have a certain folder structure. Also perhaps we should call the font-family `"material-icons"` instead

File: docs/gulpfile.js (hunk 6 above)
Comment (ThomasBurleson) [inline docs/gulpfile.js]: This means that the @programmist  code base has not been upated with Angular Material master.
This stuff should NOT be part of the PR.

File: gulpfile.js (hunk 7 above)
Comment (ThomasBurleson) [inline gulpfile.js]: Should not be part of the PR

File: src/components/icon/icon.js (hunk 8 above)
Comment (ThomasBurleson) [inline src/components/icon/icon.js]: nice :+1:

File: src/components/icon/icon.scss (hunk 10 above)
Comment (ThomasBurleson) [inline src/components/icon/icon.scss]: noice

File: docs/gulpfile.js (hunk 6 above)
Comment (programmist) [inline docs/gulpfile.js]: I did rebase from upstream master.  I'll do that again to see if it changes anything.

File: gulpfile.js (hunk 7 above)
Comment (programmist) [inline gulpfile.js]: We did make some gulpfile changes.  Should those be in a separate PR?

File: docs/gulpfile.js (hunk 6 above)
Comment (ThomasBurleson) [inline docs/gulpfile.js]: @programmist - the upstream master `docs/gulpfile.js`  has been update
